source/jsonParsing.ml

```diff
@@ -64,3 +64,201 @@ let optional_list_member ~f name json =
+    | `Float of float
+    | `Int of int
+    | `List of t list
+    | `Assoc of (string * t) list
+    ]
+
+  and t = expression Node.t [@@deriving show, eq]
+
+  exception
+    TypeError of {
+      message: string;
+      json: t;
+    }
+
+  let null_node = { Node.location = None; value = `Null }
+
+  let json_of_string input =
+    let dec d =
+      match Jsonm.decode d with
+      | `Lexeme l -> l
+      | `Error error ->
+          raise
+            (JsonmException
+               {
+                 message = Format.asprintf "%a" Jsonm.pp_error error;
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+      | `End
+      | `Await ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered end of file or input";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    in
+    let rec value v k d =
+      let location = Location.from_decoded_range (Jsonm.decoded_range d) in
+      match v with
+      | `Os -> obj [] k d
+      | `As -> arr [] k d
+      | (`Null | `Bool _ | `String _ | `Float _) as v ->
+          k { Node.location = Some location; value = v } d
+      | _ ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered unexpected token or element";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    and arr vs k d =
+      match dec d with
+      | `Ae ->
+          k
+            {
+              Node.location = Some (Location.from_decoded_range (Jsonm.decoded_range d));
+              value = `List (List.rev vs);
+            }
+            d
+      | v -> value v (fun v -> arr (v :: vs) k) d
+    and obj ms k d =
+      match dec d with
+      | `Oe ->
+          k
+            {
+              Node.location = Some (Location.from_decoded_range (Jsonm.decoded_range d));
+              value = `Assoc (List.rev ms);
+            }
+            d
+      | `Name n -> value (dec d) (fun v -> obj ((n, v) :: ms) k) d
+      | _ ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered unexpected token or element";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    in
+    let d = Jsonm.decoder (`String input) in
+    value (dec d) (fun v _ -> v) d
+
+
+  module Util = struct
+    open Core
+
+    let get_type_error expected json =
+      let get_message current = "Expected " ^ expected ^ ", got " ^ current in
+      let message =
+        match json.Node.value with
+        | `Assoc _ -> get_message "object"
+        | `Int _ -> get_message "integer"
+        | `Float _ -> get_message "float"
+        | `Null -> get_message "null"
+        | `String _ -> get_message "string"
+        | `Bool _ -> get_message "boolean"
+        | `List _ -> get_message "list"
+      in
```

Comment:
it feels a bit unnatural written this way. Why not just something like:
```
let current =
   match json.Node.value with
  | `Assoc _ -> "object"
  |...
in
Format.sprintf "Expected %s, got %s" expected current
```

Reply:
yeah, better